Add `state` param to OAuth `/authorize` request

msabramo opened the pull request:
Add `state` param to OAuth `/authorize` request; increases security and allows OAuth to work with Okta

## Motivation and Context
Some OAuth servers, such as Okta require the `state` parameter to be present to help prevent CSRF attacks.
More info:
- upstream issue - https://github.com/modelcontextprotocol/inspector/issues/442
- discussion - https://github.com/orgs/mcp-auth/discussions/44
- article from Okta - https://support.okta.com/help/s/article/the-authentication-request-has-an-invalid-state-parameter?language=en_US

In particular, https://support.okta.com/help/s/article/the-authentication-request-has-an-invalid-state-parameter?language=en_US says:
> Okta requires the OAuth 2.0 `state` parameter on all requests to the `/authorize` endpoint to prevent cross-site request forgery (CSRF). The OAuth 2.0 specification [requires](https://tools.ietf.org/html/rfc6749#section-10.12) that clients protect their redirect URIs against CSRF by sending a value in the authorized request that binds the request to the user-agent's authenticated state. Using the `state` parameter is also a countermeasure to several other known attacks, as outlined in [OAuth 2.0 Threat Model and Security Considerations](https://tools.ietf.org/html/rfc6819).

## How Has This Been Tested?
- `npm run dev`
- Connect to ["whoami" SSE server](https://mcp-auth.dev/docs/tutorials/whoami) at http://localhost:3001/sse
- using a corporate Okta instance as the authorization server.
- Was able to:
   - successfully login
   - execute a `whoami` tool
   - verify that the tool knew info about me from OIDC.
 
<img width="1800" alt="Screenshot 2025-05-31 at 12 51 51 AM" src="https://github.com/user-attachments/assets/41f3b4aa-ba1b-408b-bf38-ff512be9c175" />

## Breaking Changes
Hopefully none

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
    <img width="1229" alt="Screenshot 2025-05-31 at 12 47 23 AM" src="https://github.com/user-attachments/assets/6dba0330-524c-4e34-8666-b11a55be7e29" />
- [ ] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

Cc: @xiaoyijun, @phuctm97, @jspahrsummers, @jerome3o-anthropic

msabramo committed:
Add `state` param to OAuth /authorize request

When doing an OAuth `/authorize` request to an OAuth authorization
server, include a `state` parameter, as described at
https://support.okta.com/help/s/article/the-authentication-request-has-an-invalid-state-parameter?language=en_US
This enables OAuth to work with Okta.

Fixes: https://github.com/orgs/mcp-auth/discussions/44, modelcontextprotocol/inspector#442

msabramo committed:
npm run prettier-fix

msabramo committed:
Add `generateRandomState` test to `oauthUtils.ts`

msabramo committed:
Add client/src/lib/auth.state.test.ts

msabramo committed:
Add client/src/lib/auth.authorize-url.test.ts

xiaoyijun commented:
Hi @msabramo, thanks for your PR! I am currently considering syncing the mcp-auth fork of inspector with the official repo. Would you be open to submitting this PR to the official inspector repo as well? Of course, I will first merge this PR into the current project.

I plan to do this after https://github.com/modelcontextprotocol/inspector/pull/345 is merged. If there are further updates to the official inspector repo, would you be able to help keep things in sync there?

Previously, I was syncing PRs from my personal fork to the official repo, so submitting PRs to the official repo directly from this project (mcp-auth) is a bit inconvenient.

Thanks again for your contribution!

xiaoyijun commented on client/src/lib/auth.ts:
```diff
@@ -98,7 +99,36 @@ export class InspectorOAuthClientProvider implements OAuthClientProvider {
+   */
+  generateAndStoreState(): string {
+    const state = generateRandomState(32);
+    const key = getServerSpecificKey("oauth_state", this.serverUrl);
```
```suggestion
    const key = getServerSpecificKey(SESSION_KEYS.OAUTH_STATE, this.serverUrl);
```

msabramo replied:
Good idea! Implemented in 0de8a3634a4eb1bbaa7f99af408d8d12ecb48375.

msabramo committed:
Replace "oauth_state" with SESSION_KEYS.OAUTH_STATE

msabramo commented:
> Hi @msabramo, thanks for your PR! I am currently considering syncing the mcp-auth fork of inspector with the official repo. Would you be open to submitting this PR to the official inspector repo as well?

You're welcome! I'm happy to contribute to the very nice work you've done here to make auth much easier in MCP and I'd be happy to submit this to the official repo tool so that even more people can benefit from it!

I already have a few PRs in the official inspector repo:
https://github.com/modelcontextprotocol/inspector/pulls/msabramo

xiaoyijun commented:
Hi @msabramo, some tests failed—could you take a look?

msabramo commented:
> Hi @msabramo, some tests failed—could you take a look?

Well, it looks like these same errors might be happening on `main`, as a result of a push you did 3 weeks ago?

https://github.com/mcp-auth/inspector/actions/runs/15085688162

Since I'm looking at it now, let me see if I can fix them and submit another PR for that.

msabramo committed:
Fix a bunch of TypeScript build errors

from apparently forgetting to add some new defaultProps

```
> @modelcontextprotocol/inspector-client@0.10.2 build
> tsc -b && vite build

src/components/__tests__/Sidebar.test.tsx:44:10 - error TS2739: Type '{ connectionStatus: "disconnected"; transportType: "stdio"; setTransportType: Mock<UnknownFunction>; command: string; setCommand: Mock<UnknownFunction>; ... 16 more ...; setConfig: Mock<...>; }' is missing the following properties from type 'SidebarProps': oauthClientId, setOauthClientId, oauthParams, setOauthParams

44         <Sidebar {...defaultProps} {...props} />
            ~~~~~~~

src/components/__tests__/Sidebar.test.tsx:185:12 - error TS2739: Type '{ bearerToken: string; transportType: "sse"; connectionStatus: "disconnected"; setTransportType: Mock<UnknownFunction>; command: string; setCommand: Mock<UnknownFunction>; ... 15 more ...; setConfig: Mock<...>; }' is missing the following properties from type 'SidebarProps': oauthClientId, setOauthClientId, oauthParams, setOauthParams

185           <Sidebar
               ~~~~~~~

src/components/__tests__/Sidebar.test.tsx:215:12 - error TS2739: Type '{ bearerToken: string; transportType: "sse"; connectionStatus: "disconnected"; setTransportType: Mock<UnknownFunction>; command: string; setCommand: Mock<UnknownFunction>; ... 15 more ...; setConfig: Mock<...>; }' is missing the following properties from type 'SidebarProps': oauthClientId, setOauthClientId, oauthParams, setOauthParams

215           <Sidebar
               ~~~~~~~

src/components/__tests__/Sidebar.test.tsx:379:12 - error TS2739: Type '{ env: Record<string, string>; setEnv: Mock<UnknownFunction>; connectionStatus: "disconnected"; transportType: "stdio"; setTransportType: Mock<UnknownFunction>; ... 16 more ...; setConfig: Mock<...>; }' is missing the following properties from type 'SidebarProps': oauthClientId, setOauthClientId, oauthParams, setOauthParams

379           <Sidebar {...defaultProps} env={updatedEnv} setEnv={setEnv} />
               ~~~~~~~

src/components/__tests__/Sidebar.test.tsx:414:12 - error TS2739: Type '{ env: { NEW_KEY: string; }; connectionStatus: "disconnected"; transportType: "stdio"; setTransportType: Mock<UnknownFunction>; command: string; setCommand: Mock<UnknownFunction>; ... 15 more ...; setConfig: Mock<...>; }' is missing the following properties from type 'SidebarProps': oauthClientId, setOauthClientId, oauthParams, setOauthParams

414           <Sidebar {...defaultProps} env={{ NEW_KEY: "test_value" }} />
               ~~~~~~~

src/components/__tests__/Sidebar.test.tsx:599:12 - error TS2739: Type '{ config: InspectorConfig; setConfig: Mock<UnknownFunction>; connectionStatus: "disconnected"; transportType: "stdio"; setTransportType: Mock<...>; ... 16 more ...; loggingSupported: boolean; }' is missing the following properties from type 'SidebarProps': oauthClientId, setOauthClientId, oauthParams, setOauthParams

599           <Sidebar
               ~~~~~~~

Found 6 errors.
```

msabramo commented:
> > Hi @msabramo, some tests failed—could you take a look?
> 
> Well, it looks like these same errors might be happening on `main`, as a result of a push you did 3 weeks ago?
> 
> https://github.com/mcp-auth/inspector/actions/runs/15085688162
> 
> Since I'm looking at it now, let me see if I can fix them and submit another PR for that.

Hi @xiaoyijun, check out https://github.com/mcp-auth/inspector/pull/4

msabramo committed:
Merge branch 'fix-ts-build-errors' into send-state-parameter-for-oauth-authorize-requests

msabramo committed:
Fix type error on window.location in auth.authorize-url.test.ts
